docs/development/best-practices.md

```diff
@@ -0,0 +1,115 @@
+# Best Practices
+
+## General
+
+- Prefer full function declaration, as opposed to `const func = ():void => {}`, for readability and better stack traces
+- Avoid [Enums](https://github.com/renovatebot/renovate/issues/13743), use union
+  or [immutable objects](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/pep440/range.ts#L8-L20) instead
+- Always add unit tests for full code coverage
+  - Only use `istanbul` comments for unreachable code coverage that is needed for `codecov` completion.
+  - Use descriptive `istanbul` comments
+
+```js
+// istanbul ignore next: can never happen
+```
+
+### Logging
+
+- Use logger metadata:
+
+Use
+
+```js
+logger.trace({ obj }, 'title');
+```
+
+instead of
+
+```js
+logger.trace('title:\n' + obj);
+```
+
+### Array constructor
+
+Avoid the `Array()` constructor, with or without `new`, in your TypeScript code.
+It has confusing and contradictory usage.
+
+```js
+const a = new Array(2); // [undefined, undefined]
+const b = new Array(2, 3); // [2, 3];
+```
+
+Instead, always use bracket notation to initialize arrays, or `from` to initialize an Array with a certain size.
+
+[Source](https://google.github.io/styleguide/tsguide.html#array-constructor)
+
+### Iterating objects & containers
+
+Use `for ( ... of ...)` loops instead of `[Array|Set|Map].prototype.forEach` and `for ( ... in ...)`.
+
+- Using `for ( ... in ...)` for objects is error-prone. It will include enumerable properties from the prototype chain
+- Using `for ( ... in ...)` to iterate over arrays, will counterintuitively give the array's indices
+- Avoid `[Array|Set|Map].prototype.forEach`. It makes code harder to debug and defeats some useful compiler checks like
+  reachability
+
+Only use `Array.prototype.map()` when the return value is used, otherwise use `for ( ... of ...)`.
+
+[Source](https://google.github.io/styleguide/tsguide.html#iterating-objects)
+
+### Exports
+
+Use named exports in all code.
+Avoid default `exports`.
+This way all `imports` follow the same pattern.
+
+[Source, reasoning and examples.](https://google.github.io/styleguide/tsguide.html#exports)
+
+### Imports
+
+Use ES6 module syntax.
```

Comment:
Is it clear to people what ES6 module syntax is? Do we need to add a link to some docs?

Reply:
added a link